sw/device/tests/sim_dv/sysrst_ctrl_ulp_z3_wakeup_test.c

```diff
@@ -0,0 +1,186 @@
+ * the testbench is ready to proceed with the next phase of the test.
+ */
+static void wait_next_test_phase(void) {
+  // Set WFI status for testbench synchronization,
+  // no actual WFI instruction is issued.
+  test_status_set(kTestStatusInWfi);
+  test_status_set(kTestStatusInTest);
+  flash_ctrl_testutils_backdoor_wait_update(&flash, (uintptr_t)&kTestPhase,
+                                            kTestPhaseTimeoutUsec);
+}
+
+/**
+ * Configure *_debounce_ctl and then enable ULP wakeup.
+ */
+static void configure_wakeup(void) {
+  dif_sysrst_ctrl_ulp_wakeup_config_t wakeup_config;
+
+  // Keep toggle disabled when writing debounce configuration
+  wakeup_config.enabled = kDifToggleDisabled;
+  wakeup_config.ac_power_debounce_time_threshold = kDebounceTimer;
+  wakeup_config.lid_open_debounce_time_threshold = kDebounceTimer;
+  wakeup_config.power_button_debounce_time_threshold = kDebounceTimer;
+
+  CHECK_DIF_OK(
+      dif_sysrst_ctrl_ulp_wakeup_configure(&sysrst_ctrl, wakeup_config));
+
+  CHECK_DIF_OK(
+      dif_sysrst_ctrl_ulp_wakeup_set_enabled(&sysrst_ctrl, kDifToggleEnabled));
+}
+
+static void go_to_sleep(void) {
+  // Wakeup source is from sysrst_ctrl (source one).
+  LOG_INFO("Going to sleep.");
+  test_status_set(kTestStatusInWfi);
+  rstmgr_testutils_pre_reset(&rstmgr);
+  pwrmgr_testutils_enable_low_power(&pwrmgr, kDifPwrmgrWakeupRequestSourceOne,
+                                    /*pwrmgr_domain_config=*/0);
+  wait_for_interrupt();
+}
+
+static void check_wakeup_reason(void) {
+  dif_rstmgr_reset_info_bitfield_t rst_info;
+  rst_info = rstmgr_testutils_reason_get();
+  rstmgr_testutils_reason_clear();
+
+  CHECK(rst_info == kDifRstmgrResetInfoLowPowerExit,
+        "Wrong reset reason = %02X", rst_info);
+}
+
+static bool has_wakeup_happened(void) {
+  bool wakeup_detected;
+  CHECK_DIF_OK(
+      dif_sysrst_ctrl_ulp_wakeup_get_status(&sysrst_ctrl, &wakeup_detected));
+  return wakeup_detected;
+}
+
+bool test_main(void) {
+  CHECK_DIF_OK(dif_sysrst_ctrl_init(
+      mmio_region_from_addr(TOP_EARLGREY_SYSRST_CTRL_AON_BASE_ADDR),
+      &sysrst_ctrl));
+  CHECK_DIF_OK(dif_pinmux_init(
+      mmio_region_from_addr(TOP_EARLGREY_PINMUX_AON_BASE_ADDR), &pinmux));
+  CHECK_DIF_OK(dif_pwrmgr_init(
+      mmio_region_from_addr(TOP_EARLGREY_PWRMGR_AON_BASE_ADDR), &pwrmgr));
+  CHECK_DIF_OK(dif_rstmgr_init(
+      mmio_region_from_addr(TOP_EARLGREY_RSTMGR_AON_BASE_ADDR), &rstmgr));
+
+  flash_ctrl_testutils_backdoor_init(&flash);
+
+  while (kTestPhase < kTestPhaseDone) {
+    switch (kTestPhase) {
+      case kTestPhaseInit:
+        pinmux_setup();
+        break;
+      case kTestPhaseDriveZero:
+        configure_wakeup();
+        LOG_INFO("kTestPhaseDriveZero");
+        break;
+      case kTestPhaseWaitNoWakeup:
+        CHECK(!has_wakeup_happened());
+        LOG_INFO("kTestPhaseWaitNoWakeup");
+        go_to_sleep();
+        check_wakeup_reason();
```

Comment:
Hm I don't think this is this the right place to check for the wakeup reason.
If we are going to deep sleep, the core will be shut off and reset, meaning that we will never execute this...

It's probably better to do the following:
- modify `check_wakeup_reason()` to `reset_is_low_power_exit()` and return a bool
- add `CHECK(!reset_is_low_power_exit());` before `CHECK(!reset_is_low_power_exit());`
- add `CHECK(reset_is_low_power_exit());` before `CHECK(reset_is_low_power_exit());`